awswrangler/_utils.py

```diff
@@ -65,6 +67,58 @@
 
 Boto3PrimitivesType = Dict[str, Optional[str]]
 
+# A mapping from import name to package name (on PyPI) for packages where
+# these two names are different.
+INSTALL_MAPPING = {
+    "redshift_connector": "redshift",
+    "pymysql": "mysql",
+    "pg8000": "postgres",
+    "pyodbc": "sqlserver",
+    "gremlin_python": "gremlin",
+    "opensearchpy": "opensearch",
+}
+
+
+def check_optional_dependency(
+    module: Optional[ModuleType],
+    name: str,
+) -> Callable[..., Any]:
+    def decorator(func: Callable[..., Any]) -> Any:
+        @wraps(func)
+        def inner(*args: Any, **kwargs: Any) -> Any:
+            if not module:
+                package_name = INSTALL_MAPPING.get(name)
+                install_name = package_name if package_name is not None else name
+                raise ModuleNotFoundError(
+                    f"Missing optional dependency '{name}'. " f"Use pip awswrangler[{install_name}] to install it."
+                )
+            return func(*args, **kwargs)
+
+        return inner
+
+    return decorator
+
+
+def import_optional_dependency(name: str) -> ModuleType:
```

Comment:
According to the documentation, this should return `Optional[ModuleType]`

Reply:
So when set to `Optional[ModuleType]`, mypy is not smart enough to realise that the decorator `check_optional_dependency` protects the function from calling an attribute on `None` ([example](https://github.com/aws/aws-sdk-pandas/blob/feat/issue-1983-optional-deps/awswrangler/opensearch/_read.py#L45)) and instead it raises `union-attr` errors all over the place. Not sure how to get around it

Reply:
I see your point. I looked into and couldn't find a better way either.